File: typescript/day9_part2_2017.ts
import * as fs from 'fs';

function processStream(input: string): [number, number] {
    let score = 0;
    let totalScore = 0;
    let garbageCount = 0;
    let inGarbage = false;
    let cancelNext = false;
    let depth = 0;

    for (let i = 0; i < input.length; i++) {
        const char = input[i];

        if (cancelNext) {
            cancelNext = false;
            continue;
        }

        if (char === '!') {
            cancelNext = true;
            continue;
        }

        if (inGarbage) {
            if (char === '>') {
                inGarbage = false;
            } else {
                garbageCount++;
            }
        } else {
            if (char === '<') {
                inGarbage = true;
            } else if (char === '{') {
                depth++;
            } else if (char === '}') {
                score += depth;
                depth--;
            }
        }
    }

    return [score, garbageCount];
}

function main() {
    const input = fs.readFileSync('input.txt', 'utf8');
    const [totalScore, garbageCount] = processStream(input);

    console.log(`Total score for all groups: ${totalScore}`);
    console.log(`Non-canceled characters within the garbage: ${garbageCount}`);
}

main();

File: r/day9_part2_2017.R

input <- readLines("input.txt")
stream <- strsplit(input, "")[[1]]

score <- 0
total_score <- 0
garbage_count <- 0
ignore_next <- FALSE
in_garbage <- FALSE

for (char in stream) {
  if (ignore_next) {
    ignore_next <- FALSE
  } else if (char == "!") {
    ignore_next <- TRUE
  } else if (in_garbage) {
    if (char == ">") {
      in_garbage <- FALSE
    } else {
      garbage_count <- garbage_count + 1
    }
  } else {
    if (char == "{") {
      score <- score + 1
      total_score <- total_score + score
    } else if (char == "}") {
      score <- score - 1
    } else if (char == "<") {
      in_garbage <- TRUE
    }
  }
}

print(total_score)
print(garbage_count)
